apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaProducerHeadersInstrumentation.java

```diff
@@ -71,21 +71,21 @@ public static Object[] beforeSend(@Advice.FieldValue("apiVersions") final ApiVer
                                           @Advice.Argument(0) final ProducerRecord<?, ?> record,
                                           @Nullable @Advice.Argument(value = 1) Callback callback) {
             Span<?> span = helper.onSendStart(record);
-            if (span == null) {
-                return null;
-            }
 
             // Avoid adding headers to records sent to a version older than 0.11.0 - see specifications in
             // https://kafka.apache.org/0110/documentation.html#messageformat
             if (apiVersions.maxUsableProduceMagic() >= RecordBatch.MAGIC_VALUE_V2 && headersSupported) {
                 try {
-                    headersHelper.setOutgoingTraceContextHeaders(span, record);
+                    headersHelper.setOutgoingTraceContextHeaders(tracer.currentContext(), record);
                 } catch (final IllegalStateException e) {
                     // headers are in a read-only state
                     logger.debug("Failed to add header to Kafka record {}, probably to headers' read-only state.", record);
                 }
             }
 
+            if (span == null) {
+                return null;
+            }
```

Comment:
:tada: it looks like you've fixed a bug where the context propagation headers might not have been sent when no span was created (which is definitely a corner case).